lib/navigation.php

```diff
@@ -0,0 +1,355 @@
+	}
+
+	// Set up the $menu_item variables.
+	_wp_menu_item_classes_by_context( $menu_items );
+
+	$sorted_menu_items = array();
+	foreach ( (array) $menu_items as $menu_item ) {
+		$sorted_menu_items[ $menu_item->menu_order ] = $menu_item;
+	}
+
+	unset( $menu_items, $menu_item );
+
+	// END: Code that already exists in wp_nav_menu().
+
+	$menu_items_by_parent_id = array();
+	foreach ( $sorted_menu_items as $menu_item ) {
+		$menu_items_by_parent_id[ $menu_item->menu_item_parent ][] = $menu_item;
+	}
+
+	$navigation_block = array(
+		'blockName'   => 'core/navigation',
+		'attrs'       => array(),
+		'innerBlocks' => gutenberg_convert_menu_items_to_blocks(
+			isset( $menu_items_by_parent_id[0] )
+				? $menu_items_by_parent_id[0]
+				: array(),
+			$menu_items_by_parent_id
+		),
+	);
+
+	return render_block( $navigation_block );
+}
+add_filter( 'pre_wp_nav_menu', 'gutenberg_output_block_nav_menu', 10, 2 );
+
+/**
+ * Shim that makes nav-menus.php nicely display a menu item with its `type` set to
+ * `'block'`.
+ *
+ * Specifically, this shim makes it so that `Walker_Nav_Menu_Edit::start_el()`
+ * outputs extra form fields. When merged to Core, this markup should exist in
+ * `Walker_Nav_Menu_Edit::start_el()`.
+ *
+ * This shim can be removed when the Gutenberg plugin requires a WordPress
+ * version that has the ticket below.
+ *
+ * @see https://core.trac.wordpress.org/ticket/50544
+ *
+ * @param int     $item_id Menu item ID.
+ * @param WP_Post $item    Menu item data object.
+ */
+function gutenberg_output_block_menu_item_custom_fields( $item_id, $item ) {
+	if ( 'block' === $item->type ) {
+		?>
+		<p class="description description-wide">
+			<textarea readonly><?php echo esc_textarea( trim( $item->content ) ); ?></textarea>
+		</p>
+		<?php
+	}
+}
+add_action( 'wp_nav_menu_item_custom_fields', 'gutenberg_output_block_menu_item_custom_fields', 10, 2 );
+
+/**
+ * Shim that adds extra styling to nav-menus.php. This lets us style menu items
+ * that have a `type` set to `'block'`. When merged to Core, this CSS should be
+ * moved to nav-menus.css.
+ *
+ * This shim can be removed when the Gutenberg plugin requires a WordPress
+ * version that has the ticket below.
+ *
+ * @see https://core.trac.wordpress.org/ticket/50544
+ *
+ * @param string $hook The current admin page.
+ */
+function gutenberg_add_block_menu_item_styles_to_nav_menus( $hook ) {
+	if ( 'nav-menus.php' === $hook ) {
+		$css = <<<CSS
+			/**
+			 * HACK: We're hiding the description field using CSS because this
+			 * cannot be done using a filter. When merged to Core, we should
+			 * actually remove the field from
+			 * `Walker_Nav_Menu_Edit::start_el()`.
+			 */
+			.menu-item-block .description:first-child {
+				display: none;
+			}
+
+			.menu-item-block textarea {
+				width: 100%;
+				min-height: 100px;
+			}
+CSS;
+		wp_add_inline_style( 'nav-menus', $css );
+	}
+}
```

Comment:
This doesn't take into account that the user might have other fields enabled in screen options.

Might be best to add a unique class to the custom field wrapper, and then use `:not` to hide all the others.

Having said that, I wonder if we can get away with just CSS hacks. An issue I spotted is that it's possible to nest blocks in the nav-menus screen in a way that's incorrect (e.g. adding a link to search) ...

Reply:
>  it's possible to nest blocks in the nav-menus screen in a way that's incorrect (e.g. adding a link to search) .
That is something. If they're menu items I am not sure we can make them un-nestable.

Reply:
> Might be best to add a unique class to the custom field wrapper, and then use :not to hide all the others.

Done in 641aebb17b. It relies on extra fields having the `.description` class which isn't necessarily the case for e.g. third party plugins. I think it's the best we can do without patching Core.

> Having said that, I wonder if we can get away with just CSS hacks. An issue I spotted is that it's possible to nest blocks in the nav-menus screen in a way that's incorrect (e.g. adding a link to search) ...

That's an interesting point... maybe we could disable re-ordering Block menu items? I'm not sure.

Reply:
I wonder if we should not allow using this UI at all if the menu contains blocks.

Is there a purpose to have multiple interfaces for the same thing?

Reply:
We'll need to support disabling the new screen and functionality entirely (e.g. using the Classic Editor plugin) because there will be some users that prefer the old screen and some setups that rely on the `nav-menus.php` DOM structure. 

If the new screen and functionality is disabled, I don't want any blocks that were inserted into a menu to go "missing". Hence the need to display them in `nav-menus.php`.